remix-solidity/src/compiler/compiler.ts

```diff
@@ -0,0 +1,392 @@
+    this.state.evmVersion = _evmVersion
+  }
+
+  setLanguage (_language: Language) {
+    this.state.language = _language
+  }
+
+  internalCompile (files: Source, missingInputs?: any[]) {
+    this.gatherImports(files, missingInputs, (error, input) => {
+      if (error) {
+        this.state.lastCompilationResult = null
+        this.event.trigger('compilationFinished', [false, {'error': { formattedMessage: error, severity: 'error' }}, files])
+      } else {
+        this.state.compileJSON(input)
+      }
+    })
+  }
+
+  compile (files: Source, target: string) {
+    this.state.target = target
+    this.event.trigger('compilationStarted', [])
+    this.internalCompile(files)
+  }
+
+  setCompileJSON (_compileJSON) {
+    this.state.compileJSON = _compileJSON
+  }
+
+  onCompilerLoaded (version: string) {
+    this.state.currentVersion = version
+    this.event.trigger('compilerLoaded', [version])
+  }
+
+  onInternalCompilerLoaded () {
+    if (this.state.worker === null) {
+      let compiler
+      if (typeof (window) === 'undefined') {
+        compiler = require('solc')
+      } else {
+        compiler = solc(window['Module'])
+      }
+
+      this.state.compileJSON = (source: SourceWithTarget) => {
+        let missingInputs: any = []
+        let missingInputsCallback = (path) => {
+          missingInputs.push(path)
+          return { error: 'Deferred import' }
+        }
+
+        let result
+        try {
+          if(source && source.sources) {
+            let input = compilerInput(source.sources, {optimize: this.state.optimize, evmVersion: this.state.evmVersion, language: this.state.language})
+            result = compiler.compile(input, { import: missingInputsCallback })
+            result = JSON.parse(result)
+          }
+        } catch (exception) {
+          result = { error: { formattedMessage: 'Uncaught JavaScript exception:\n' + exception, severity: 'error', mode: 'panic' } }
+        }
+        this.compilationFinished(result, missingInputs, source)
+      }
+      this.onCompilerLoaded(compiler.version())
+    }
+  }
+ 
+  /**
+    * return the contract obj of the given @arg name. Uses last compilation result.
+    * return null if not found
+    * @param {String} name    - contract name
+    * @returns contract obj and associated file: { contract, file } or null
+    */
+  getContract (name: string) {
+    if (this.state.lastCompilationResult && this.state.lastCompilationResult.data && this.state.lastCompilationResult.data.contracts) {
+      return txHelper.getContract(name, this.state.lastCompilationResult.data.contracts)
+    }
+    return null
+  }
+
+  /**
+    * call the given @arg cb (function) for all the contracts. Uses last compilation result
+    * @param {Function} cb    - callback
+    */
+  visitContracts (cb: any) {
+    if (this.state.lastCompilationResult && this.state.lastCompilationResult.data && this.state.lastCompilationResult.data.contracts) {
```

Comment:
Normally you can just do :
```typescript
return this.state.lastCompilationResult && this.state.lastCompilationResult.data && this.state.lastCompilationResult.data.contracts
```
it will return undefined if one of them is false